storage/src/journal/variable.rs

```diff
@@ -1,2223 +1,1443 @@
-            tracked,
-            synced,
-            pruned,
-
-            _phantom: PhantomData,
+            data,
+            offsets,
+            items_per_section,
+            size,
+            oldest_retained_pos,
         })
     }
 
-    /// Ensures that a section pruned during the current execution is not accessed.
-    fn prune_guard(&self, section: u64) -> Result<(), Error> {
-        if section < self.oldest_retained_section {
-            Err(Error::AlreadyPrunedToSection(self.oldest_retained_section))
-        } else {
-            Ok(())
+    /// Rewind the journal to the given size, discarding items from the end.
+    ///
+    /// After rewinding to size N, the journal will contain exactly N items,
+    /// and the next append will receive position N.
+    ///
+    /// # Errors
+    ///
+    /// Returns [Error::InvalidRewind] if size is invalid (too large or points to pruned data).
+    /// Returns an error if the underlying storage operation fails.
+    ///
+    /// Errors may leave the journal in an inconsistent state. The journal should be closed and
+    /// reopened to trigger repair in [Journal::init].
+    pub async fn rewind(&mut self, size: u64) -> Result<(), Error> {
+        // Validate rewind target
+        match size.cmp(&self.size) {
+            std::cmp::Ordering::Greater => return Err(Error::InvalidRewind(size)),
+            std::cmp::Ordering::Equal => return Ok(()), // No-op
+            std::cmp::Ordering::Less => {}
         }
-    }
 
-    /// Reads an item from the blob at the given offset.
-    pub(crate) async fn read(
-        compressed: bool,
-        cfg: &V::Cfg,
-        blob: &Append<E::Blob>,
-        offset: u32,
-    ) -> Result<(u32, u32, V), Error> {
-        // Read item size
-        let mut hasher = crc32fast::Hasher::new();
-        let offset = offset as u64 * ITEM_ALIGNMENT;
-        let size = blob.read_at(vec![0; 4], offset).await?;
-        hasher.update(size.as_ref());
-        let size = u32::from_be_bytes(size.as_ref().try_into().unwrap()) as usize;
-        let offset = offset.checked_add(4).ok_or(Error::OffsetOverflow)?;
-
-        // Read remaining
-        let buf_size = size.checked_add(4).ok_or(Error::OffsetOverflow)?;
-        let buf = blob.read_at(vec![0u8; buf_size], offset).await?;
-        let buf = buf.as_ref();
-        let offset = offset
-            .checked_add(buf_size as u64)
-            .ok_or(Error::OffsetOverflow)?;
-
-        // Read item
-        let item = &buf[..size];
-        hasher.update(item);
-
-        // Verify integrity
-        let checksum = hasher.finalize();
-        let stored_checksum = u32::from_be_bytes(buf[size..].try_into().unwrap());
-        if checksum != stored_checksum {
-            return Err(Error::ChecksumMismatch(stored_checksum, checksum));
+        // Rewind never updates oldest_retained_pos.
+        if size < self.oldest_retained_pos {
+            return Err(Error::InvalidRewind(size));
```

Comment:
Elsewhere we return "ItemPruned" error for indices that precede oldest_retained_pos.  No strong opinion on which is better though.